Sources/NIO/SelectorGeneric.swift

```diff
@@ -0,0 +1,376 @@
+}
+
+extension Optional {
+    internal func withUnsafeOptionalPointer<T>(_ body: (UnsafePointer<Wrapped>?) throws -> T) rethrows -> T {
+        if var this = self {
+            return try withUnsafePointer(to: &this) { x in
+                try body(x)
+            }
+        } else {
+            return try body(nil)
+        }
+    }
+}
+
+extension timespec {
+    init(timeAmount amount: TimeAmount) {
+        let nsecPerSec: Int64 = 1_000_000_000
+        let ns = amount.nanoseconds
+        let sec = ns / nsecPerSec
+        self = timespec(tv_sec: Int(sec), tv_nsec: Int(ns - sec * nsecPerSec))
+    }
+}
+
+/// Represents IO events NIO might be interested in. `SelectorEventSet` is used for two purposes:
+///  1. To express interest in a given event set and
+///  2. for notifications about an IO event set that has occured.
+///
+/// For example, if you were interested in reading and writing data from/to a socket and also obviously if the socket
+/// receives a connection reset, express interest with `[.read, .write, .reset]`.
+/// If then suddenly the socket becomes both readable and writable, the eventing mechanism will tell you about that
+/// fact using `[.read, .write]`.
+struct SelectorEventSet: OptionSet, Equatable {
+
+    typealias RawValue = UInt8
+
+    let rawValue: RawValue
+
+    /// It's impossible to actually register for no events, therefore `_none` should only be used to bootstrap a set
+    /// of flags or to compare against spurious wakeups.
+    static let _none = SelectorEventSet([])
+
+    /// Connection reset or other errors.
+    static let reset = SelectorEventSet(rawValue: 1 << 0)
+
+    /// EOF at the read/input end of a `Selectable`.
+    static let readEOF = SelectorEventSet(rawValue: 1 << 1)
+
+    /// Interest in/availability of data to be read
+    static let read = SelectorEventSet(rawValue: 1 << 2)
+
+    /// Interest in/availability of data to be written
+    static let write = SelectorEventSet(rawValue: 1 << 3)
+
+    /// EOF at the write/output end of a `Selectable`.
+    ///
+    /// - note: This is rarely used because in many cases, there is no signal that this happened.
+    static let writeEOF = SelectorEventSet(rawValue: 1 << 4)
+
+    init(rawValue: SelectorEventSet.RawValue) {
+        self.rawValue = rawValue
+    }
+}
+
+internal let isEarlyEOFDeliveryWorkingOnThisOS: Bool = {
+    #if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
+    return false // rdar://53656794 , once fixed we need to do an OS version check here.
+    #else
+    return true
+    #endif
+}()
+
+/// This protocol defines the methods that are expected to be found on
+/// `Selector`. While defined as a protocol there is no expectation that any
+/// object other than `Selector` will implement this protocol: instead, this
+/// protocol acts as a reference for what new supported selector backends
+/// must implement.
+protocol _SelectorBackendProtocol {
+    associatedtype R : Registration
+    func _init() throws
```

Comment:
I'd probably give `_init` and `_deinit` better names for what they're meant to do. Maybe `initialiseState` and `validateAssertionsOnDeinit` or something? Dunno, not super important, I'll leave this up to you.

Reply:
Sure, updating.